---
title: "ML_Metagenome_GB21"
author: "Sydney Salley, Maggie Weng, Patrick Thieringer"
date: "7/2/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Genome Data-Mining
### Download genomes from NCBI genbank + refseq that were then placed in the directories "genbank" and "refseq"

```{bash, eval = F}
ncbi-genome-download -g cyanobium bacteria
ncbi-genome-download -g cyanobium -s genbank bacteria
```

### Preprocessing from genbank files to fasta + gene calls + function using anvio-script-process-genbank
Ran in both genbank and refseq folders to capture all the files
Generated the 00_LOGS_pan_workflow directory 
```{bash, eval=F}
for prefix in $(ls) 
do anvi-script-process-genbank -i $prefix/*.gz -O $prefix
done
```

### Creating the cyanobium_fa.txt file with names and paths of all relevant files for pangenome workflow
```{bash, eval=F}
ls */bacteria/*.fa  | awk -F "-" 'BEGIN {print "name\tpath\texternal_gene_calls\tgene_functional_annotation"}{print $1 "\t" $0 "\t" $1 "-external-gene-calls.txt\t" $1 "-external-functions.txt"}' | sed "s,^[a-z/]*/,," > cyanobium_fa.txt
```

### Removing junky genomes that had no gene calls
```{bash, eval=F}
grep "contains characters that are not" 00_LOGS_pan_workflow/*-anvi_gen_contigs_database.log
```

```{bash, eval=F}
Output:
GCA_002737505.1
GCA_015207535.1
GCA_015207615.1
GCF_015207535.1
GCF_015207615.1
```
Removed in vim using dd

### Creating a .json file
config_pangenomics.json file (added in 'ignore-internal-stop-codons' flag)
```{bash, eval=F}
{
    "workflow_name": "pangenomics",
    "config_version": "2",
    "project_name": "CYANOBIUM-NCBI",
    "external_genomes": "cyanobium-genbank30-external-genomes.txt",
    "fasta_txt": "cyanobium_fa.txt",
    "max_threads": 2,
    "anvi_gen_contigs_database": {
        "--ignore-internal-stop-codons": true
    },
    "output_dirs": {
        "FASTA_DIR": "01_FASTA_contigs_workflow",
        "CONTIGS_DIR": "02_CONTIGS_contigs_workflow",
        "LOGS_DIR": "00_LOGS_pan_workflow"
    }
}

```

### Editing our gene-calls.txt file
Anvio doesn't like using PGAP instead of prodigal 
But we do!
So we need to convince Anvio to accept our PGAP gene calls from NCBI genomes, which we did using DAN'S BIG BOY HACK:

```{bash, eval=F}
sed -i"" 's/NCBI_PGAP/prodigal/; s/v4.6/v2.6.3/' */bacteria/*gene-calls.txt
```

Then we reran the first part of the pangenomics workflow and exited at pangenome step (we want to include MAGS in our pangenome):
This generated the "02_CONTIGS_contigs_workflow" folder and the "cyanobium-genbank30-external-genomes.txt" file

```{bash, eval=F}
anvi-run-workflow -w pangenomics -c config_pangenomics.json
```

### Creating hmms (Hidden Markov Model) for single copy genes:
```{bash, eval=F}
for contigs in $(ls 02*/*.db) ; do anvi-run-hmms -c $contigs -T 4; done 
```

Patch because MAGs used different version of COG (14) than our genomes:
(generated the file mags_genecalls.txt)
```{bash, eval=F}
for db in $(cut -f 5 dspeth_mags_external_genome.txt | tail -2 ); do anvi-run-ncbi-cogs -c $db -T 4; done
```

### Creating a combined CYANOBIUM-GENOMES.db file with MAGs and genomes
```{bash, eval=F}
anvi-gen-genomes-storage -i dspeth_mags_external_genome.txt -e cyanobium-genbank30-external-genomes.txt -o CYANOBIUM_GENOMES.db
```

### Running anvi-pan-genomes
Generated the CYANOBIUM_PAN project directory
```{bash, eval=F}
anvi-pan-genome -g CYANOBIUM_GENOMES.db --project-name Cyanobium_Pan --output-dir “cyanobium_pan” --num-threads 4 --minbit 0.5 --mcl-inflation 6 --use-ncbi-blast
```

### Accessing the Interactive Server
We had to create ports for the Ocean server and log in with them
Dan created these ports for us
```{bash, eval=F}
ssh -L 8080:localhost:8400 yourname@ocean.caltech.edu #change 'yourname' to  to the username generated from Caltech
```
<<<<<<< HEAD
Sydney: 8455

Displaying the pangenome:
=======
>>>>>>> 938775640687806dd16f83de3a1501e9420f2385

## Displaying the pangenome:
```{bash, eval=F}
anvi-display-pan -p CYANOBIUM-PAN/Cyanobium_Pan-PAN.db -g CYANOBIUM_GENOMES.db
```

### Genomes we decided to cull after looking at stats on the interactive pangenome:
```{bash,  eval=F}
GCA2691945: low total length, low completion
GCA002714405: low total length, low completion
GCA002687115: low completion, high singletons
GCA002701375: very high singletons, low-ish completion
GCA002700895: low completion, low total length
GCA00278955: low completion, low total length
GCA01419165: low completion, missing many core gene clusters
```

### Creating a new pangenome with culled genomes:
1) created a new textfile in vim called edited_cyanobium_fa.txt from cyanobium_fa.txt with the above genomes removed completely 
2) created a new edited_config_pangenomics.json file 
3) created a new cyanobium-edited-external-genomes.txt
(to find and delete in nano, ctrl-w, type pattern, ctrl-k to delete the line containing the pattern. To do the same thing in vim, sacrifice your firstborn child and weep). 
```{bash, eval=F}
#edited_config_pangenomics.json:#

{
    "workflow_name": "pangenomics",
    "config_version": "2",
    "project_name": "CYANOBIUM-NCBI-EDIT",
    "external_genomes": "cyanobium-edited-external-genomes.txt",
    "fasta_txt": "edited_cyanobium_fa.txt",
    "max_threads": 2,
    "anvi_gen_contigs_database": {
        "--ignore-internal-stop-codons": true
    },
    "output_dirs": {
        "FASTA_DIR": "01_FASTA_contigs_workflow_edited",
        "CONTIGS_DIR": "02_CONTIGS_contigs_workflow_edited",
        "LOGS_DIR": "00_LOGS_pan_workflow_edited"
    }
}


```


### Creating a new combined EDITED-CYANOBIUM-GENOMES.db file with MAGs and reduced genomes:
```{bash, eval=F}
anvi-gen-genomes-storage -i dspeth_mags_external_genome.txt -e cyanobium-edited-external-genomes.txt -o EDITED-CYANOBIUM_GENOMES.db
```

### Rerunning anvi-pan-genomes without crap genomes:
Generated the EDITED_CYANOBIUM_PAN project directory
```{bash, eval=F}
anvi-pan-genome -g EDITED-CYANOBIUM_GENOMES.db --project-name Edited_Cyanobium_Pan --output-dir “edited_cyanobium_pan” --num-threads 4 --minbit 0.5 --mcl-inflation 6 --use-ncbi-blast
```

Displaying the pangenome after the edited version
```{bash, eval=F}
anvi-display-pan -p EDITED-CYANOBIUM-PAN/Edited_Cyanobium_Pan-PAN.db -g EDITED-CYANOBIUM_GENOMES.db
```


## GToTree

###Creating a phylogenetic tree of cyanobium genomes + MAGS using GToTree:
Creating a list of paths to contig fastas we will use:
```{bash, eval=F}
cut -f 2 edited_cyanobium_fa.txt | grep -v path > gtotree_fasta_files.txt
```

Creating a list of names that gtotree will use
```{bash, eval=F}
cut -f 1 edited_cyanobium_fa.txt | grep -v name > gtotree_accessions.txt
```

Unwrapping wrapped MAG fasta files
```{bash, eval=F}
sed ‘/>/s/^/@/; />/s/$/@/’ cyanobium_MAGs-contigs.fa | tr -d ‘\n’ | tr ‘@’ ‘\n’ | grep -v ‘^$’ > cyanobium_MAGs-contigs_unwrapped.fa
```

Separating MAGs into different files
```{bash, eval=F}
grep -A 1 MLW cyanobium_MAGs-contigs_unwrapped.fa > cyanobium_MLW_MAG-contigs.fa
grep -A 1 MLSC cyanobium_MAGs-contigs_unwrapped.fa > cyanobium_MLSC_MAG-contigs.fa
```

Making gtotree have mags too
```{bash, eval=F}
ls | grep MAG- >> mag_files.txt
cat gtotree_fasta_files.txt mag_files.txt > gtotree_fastas_mags.txt
```

This creates a phylogenetic tree based on Cyanobacteria HMM's
```{bash, eval=F}
GToTree -f gtotree_fastas_mags.txt -H Cyanobacteria -t -L Species,Strain -j 4 -o cyanobium_gtotree
```

Link to GToTree Github page: https://github.com/AstrobioMike/GToTree/wiki/example-usage
To visulaize the phylogenetic  tree: https://itol.embl.de/

This is creating a text file to put the phylogenetic tree that we created with GToTree in a format that anvio can read
```{bash, eval=F}
vi cyanobium_external_tree.txt
```

#Looking at the MAGs contigs-level mapping as an interactive view
```{bash, eval=F}
anvi-interactive -c cyanobium_MAGs-CONTIGS.db -p cyanobium_MAGs-MERGED/PROFILE.db
```

#Looking at the MAG gene-level mapping as an interactive view ordered by synteny (comparing 2 genes in order)
Anvio uses synteny to put the genes in a sequential order (preserve gene order in the contig)
We don't know if contig 2 follows contig 1, but we know that gene 2 follows gene 1 within a contig
Inner axis spokes are individual genes in the order they appeared in the contig
Therefore you can visualize groups of genes present/absent in different metagenomes in a preserved order (looking at genes and what is downstream of them)
Now we can look at blocks of differences--segments of the genome all changes together or multiple diff places across the genome? Use this to understand population differences through time


Ordered based on coverage: spokes=coverage of a gene across the metagenomes
Dendrogram groupings=similar pattern of rareness unlike the rest of the genes 
We can look at a gradient of gene rarity through depth
Gradual change implies something a bit more informative than stochastic changes in gene abundance throughout sediment depth likely due to the randomness of what's getting sedimented 
Can even go below the gene level and look at differential abundance of SNVs

Huntin' for ancient genes! (What genes are more abundant in sediment thru time, not in water column)


```{bash, eval=F}
anvi-interactive --gene-mode -c cyanobium_MAGs-CONTIGS.db -p cyanobium_MAGs-MERGED/PROFILE.db -C MAGs -b MLSC1
```


# Identifying Gene Clusters of Interest

GCA_007694915 is the most closely related genome from gene similarity to the water and sediment metagenomes.

Similarity dendrogram clades (crated a pangenome_groups.tsv file with these) 

Clade A:
GCA_000708525
GCA_009834675
GCF_009834675
GCA_000316515
GCF_000316515

Clade B: 
GCA_014191755
GCA_003249045
GCA_003248995
GCA_003249115

Clade C:
GCA_018399015
GCA_007694915
MLSC_0518_96_1
MLW_09_17_24_1

Clade D: 
GCA_014280235
GCF_014280235
GCA_900088535
GCF_900088535
GCA_000155635
GCF_000155635

Clade E:
GCA_014191535
GCA_014697415
GCF_014697415
GCA_003011885
GCF_003011885

###Looking at enrichment of COG20_PATHWAY within similar groups in the pangenome 

```{bash, eval=F}
anvi-compute-functional-enrichment -p EDITED-CYANOBIUM-PAN/Edited_Cyanobium_Pan-PAN.db -g EDITED-CYANOBIUM_GENOMES.db --category group --annotation-source COG20_PATHWAY -o cyano_pan_enriched_pathway_similarity.txt --functional-occurrence-table-output cyano-pan-pathway-frequency.txt
```

##Same thing, but enrichment of COG20_FUNCTION

```{bash, eval=F}
anvi-compute-functional-enrichment -p EDITED-CYANOBIUM-PAN/Edited_Cyanobium_Pan-PAN.db -g EDITED-CYANOBIUM_GENOMES.db --category group --annotation-source COG20_FUNCTION -o cyano_pan_enriched_function_similarity.txt --functional-occurrence-table-output cyano-pan-funcion-frequency.txt
```


Got this warning due to small group sizes:
WARNING
==============================================================
Some of your groups have very few genomes in them, so if you are running
functional enrichment, the statistical test may not be very reliable. The
minimal number of genomes in a group for the test to be reliable depends on a
number of factors, but we recommend proceeding with great caution because the
following groups have fewer than 8 genomes: D, E, B, C, A.


###Inspecting MAGs 

```{bash, eval=F}
anvi-interactive --gene-mode -c cyanobium_MAGs-CONTIGS.db -p cyanobium_MAGs-MERGED/PROFILE.db -C MAGs -b MLSC1
```

###Exporting gene coverage and detection table (run command from ml_metagenomes/dspeth_cyano_bins)

```{bash, eval=F}
anvi-export-gene-coverage-and-detection -p cyanobium_MAGs-MERGED/PROFILE.db -c cyanobium_MAGs-CONTIGS.db -O cyanobium_MAGs
```

Output files: 
cyanobium_MAGs-GENE-DETECTION.txt
cyanobium_MAGs-GENE-COVERAGES.txt

```{bash, eval=F}


## Generally Useful commands: 
```{}
(put this in a seperate screen to run:
screen -r to resume active session
ctrl-a d to detach from a session
ctrl-a k to kill a session
screen to make new session)

cat cyanobium_gtotree.tre | pbcopy # print contents of a file and send it to your copy/paste board

`top` to see what is running in a bash session 
```
```{r}

```

